### DISCUSSION TITLE
How to do context stuffing?

### DISCUSSION THREAD

**Discussion by robertvazan** [open-webui/open-webui#7447] [Q&A]
I would like to ask for your opinion, experience, and best practices regarding context stuffing and its integration in WebUI.

Context stuffing (or prompt stuffing) is the practice of filling a large context window with relevant information. I am aware of the dangers ([lost in the middle](https://arxiv.org/abs/2307.03172), [RULER](https://github.com/NVIDIA/RULER), [attention noise](https://arxiv.org/abs/2410.05258)), but my experiments with manually constructed large contexts show that context stuffing is highly effective, especially when the model is provided with several recent real-world input-output examples. I am now looking for a way to automate it.

I have considered several options:
- one long conversation, building the context naturally (messy and laborious)
- long system prompt (messy, not prompt-specific,  no impersonation)
- built-in RAG (hopelessly ineffective and buggy)
- complex external RAG (complicated, likely not that much better)
- have a script generate the context, then paste it in WebUI (messy, no impersonation)
- filters (fast-evolving, undocumented API)
- Ollama/OpenAI protocol proxy (complicated implementation, evolving protocols)

Conversation archival is a big problem with many of these approaches. I need chat history to populate the context with past examples (and perhaps to later do fine-tuning too). Stuffed context must be archived, but I also have to be able to separate it from the actual user prompt. Corrections performed by impersonation are not sent to filters nor over Ollama/OpenAI protocol. I have to scan conversation history via WebUI API and maintain separate history storage.

There seems to be no ideal solution. Is anyone else doing non-trivial context stuffing? Which integration approach are you using? How is it working for you?

**Comment by TomLucidor** [comment]
Start from
- looking into long-context models (and associated settings/params)
- be very selective to what information to pack for the job
- learn information compression and summary techniques
- use AutoPrompt and similar tools to enhance prompt based on use case

What other constraints are there?

**Comment by robertvazan** [comment]
Thanks @TomLucidor. I have meantime proceeded to implement my own custom context stuffing code. I expose it to WebUI via Ollama proxy as a virtual model. My code maintains its own separate chat, feedback, and knowledge archive that can be used for evals and fine-tuning. I have custom content compression logic, but I hope to ditch that eventually when contexts grow long enough. This solution works, but it's far more code than anticipated. I should probably opensource it in hope it saves time for others.
